File: src/main/java/july_2025/BuildTree.java
package july_2025;

import java.util.Arrays;

public class BuildTree {

    public static void main(String[] args) {
        int[] preorder = {3,9,20,15,7};
        int[] inorder = {9,3,15,20,7};

        BuildTree buildTree = new BuildTree();
        TreeNode treeNode = buildTree.buildTree(preorder, inorder);
        System.out.println(treeNode);
    }

    // O(n) time | O(n) space
    int idx = 0;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (idx == preorder.length || inorder.length == 0) {
            return null;
        }

        int nodeVal = preorder[idx];
        TreeNode newNode = new TreeNode(nodeVal);
        int i = 0;
        while (inorder[i] != nodeVal) {
            i++;
        }
        int[] leftIn = Arrays.copyOfRange(inorder, 0, i);
        int[] rightIn = Arrays.copyOfRange(inorder, i + 1, inorder.length);

        idx++;
        newNode.left = buildTree(preorder, leftIn);
        newNode.right = buildTree(preorder, rightIn);

        return newNode;
    }

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        public TreeNode(int val) {
            this.val = val;
        }
    }

}

File: python/second/build_tree.py
class Solution(object):
    # O(n) time | O(n) space
    idx = 0

    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: Optional[TreeNode]
        """
        if self.idx == len(preorder) or len(inorder) == 0:
            return None

        val = preorder[self.idx]
        i = 0
        while i < len(inorder) and inorder[i] != val:
            i += 1
        
        newNode = TreeNode(val)

        leftIn = inorder[0:i]
        rightIn = inorder[i + 1:]
        self.idx += 1
        newNode.left = self.buildTree(preorder, leftIn)
        newNode.right = self.buildTree(preorder, rightIn)

        return newNode


        
class TreeNode(object):
    def __init__(self, val, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right